### PR TITLE
feat(monitor): add perp support to monitor

### PR DESCRIPTION
**Pull request by chrismaree**
The disputer bot needs to work with the perpetual.

Obligatory issue: #2503

**Summary**

This PR follows a very similar pattern to that defined in PR https://github.com/UMAprotocol/protocol/pull/2394 and https://github.com/UMAprotocol/protocol/pull/2453 which added perpetual support to the liquidator bot and disputer bots respectively. This PR incorporated the latest changes that generalised the multi-version testing, merged in recently in PR https://github.com/UMAprotocol/protocol/pull/2448.

This PR modifies the monitor bot and all associated modeles and tests to work with the perpetual. The `BalanceMonitor.js` module was not changed as no additional functionality is needed to be compatible for this as it simply monitors token balances within an EOA.

**Details**

Primary places where the non-standard changes were implemented have been highlighted within the PR.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [X]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [X]  Existing tests adequate, no new tests required
- [X]  All existing tests pass
- [ ]  Untested

**Issue(s)**

close :#2503

### CODE REVIEW

File: packages/financial-templates-lib/src/clients/ExpiringMultiPartyClient.js [hunk 1]
```diff
@@ -293,6 +293,4 @@ class ExpiringMultiPartyClient {
   }
 }
 
-module.exports = {
-  ExpiringMultiPartyClient
-};
+module.exports = { ExpiringMultiPartyClient };
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/clients/ExpiringMultiPartyClient.js:296]: formatting nit.

File: packages/financial-templates-lib/src/clients/ExpiringMultiPartyEventClient.js [hunk 2]
```diff
@@ -40,6 +49,22 @@ class ExpiringMultiPartyEventClient {
     // Last block number to end the searching for events at.
     this.lastBlockToSearchUntil = endingBlockNumber;
     this.lastUpdateTimestamp = 0;
+
+    if (contractType !== "ExpiringMultiParty" && contractType !== "Perpetual")
+      throw new Error(
+        `Invalid contract type provided: ${contractType}! The financial product client only supports ExpiringMultiParty or Perpetual`
+      );
+    this.contractType = contractType;
+    if (
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/clients/ExpiringMultiPartyEventClient.js:58]: note: we will improve this in the future with a better version comparison logic.

File: packages/financial-templates-lib/test/clients/ExpiringMultiPartyEventClient.js [hunk 3]
```diff
@@ -1,29 +1,78 @@
 const { toWei, utf8ToHex, padRight } = web3.utils;
 const winston = require("winston");
 
-const { interfaceName, parseFixed, MAX_UINT_VAL, ZERO_ADDRESS, advanceBlockAndSetTime } = require("@uma/common");
```
Comment (chrismaree) [inline packages/financial-templates-lib/test/clients/ExpiringMultiPartyEventClient.js:4]: changes here are again very similar to other PRs adding perp support. the main logic is multi-version testing.

File: packages/liquidator/index.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ require("dotenv").config();
 const retry = require("async-retry");
 
 // Helpers
-const { MAX_UINT_VAL, findContractVersion, SUPPORTED_CONTRACT_VERSIONS } = require("@uma/common");
+const { getWeb3, MAX_UINT_VAL, findContractVersion, SUPPORTED_CONTRACT_VERSIONS } = require("@uma/common");
```
Comment (chrismaree) [inline packages/liquidator/index.js:7]: this import was repeated.

File: packages/liquidator/test/index.js [hunk 5]
```diff
@@ -418,10 +418,16 @@ contract("index.js", function(accounts) {
         // To verify decimal detection is correct for a standard feed, check the third log to see it matches expected.
         assert.isTrue(spyLogIncludes(spy, 3, `"contractVersion":"${contractVersion.contractVersion}"`));
         assert.isTrue(spyLogIncludes(spy, 3, `"contractType":"${contractVersion.contractType}"`));
-
+      });
+      it("Correctly rejects unknown contract types", async function() {
```
Comment (chrismaree) [inline packages/liquidator/test/index.js:422]: I split this test into two parts to be more explicit as to what is being tested & validate the error string correctly.

File: packages/monitors/test/CRMonitor.js [hunk 6]
```diff
@@ -35,373 +32,498 @@ const configs = [
...
+            assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${monitoredTrader}`)); // liquidator address
+            assert.isTrue(lastSpyLogIncludes(spy, `https://etherscan.io/address/${monitoredTrader}`)); // liquidator address
+            assert.isTrue(lastSpyLogIncludes(spy, "175.00%")); // calculated CR ratio for this position
+            assert.isTrue(lastSpyLogIncludes(spy, "200%")); // calculated CR ratio threshold for this address
+            assert.isTrue(lastSpyLogIncludes(spy, "1.00")); // Current price of the identifer
+            assert.isTrue(lastSpyLogIncludes(spy, hexToUtf8(await emp.priceIdentifier()))); // Synthetic identifier
+          }
+        );
+        versionedIt([{ contractType: "Perpetual", contractVersion: "latest" }])(
+          "Winston correctly emits collateralization ratio message considering perpetual funding rates",
```
Comment (chrismaree) [inline packages/monitors/test/CRMonitor.js:386]: This is a funding rate and perpetual specific test that validates that funding rates are correctly considered within the CR monitor.

**Review by daywiss — APPROVED** [review]
this is quite a feat, between this and all the other updates. you should take a vacation after this 😄

**Review by pemulis — APPROVED** [review]

File: packages/financial-templates-lib/test/clients/ExpiringMultiPartyEventClient.js [hunk 7]
```diff
@@ -1,29 +1,78 @@
...
-
+// Run the tests against 3 different kinds of token/synth decimal combinations:
+// 1) matching 18 & 18 for collateral for most token types with normal tokens.
+// 2) non-matching 8 collateral & 18 synthetic for legacy UMA synthetics.
+// 3) matching 8 collateral & 8 synthetic for current UMA synthetics.
 const configs = [
-  { tokenName: "Wrapped Ether", tokenSymbol: "WETH", collateralDecimals: 18 },
-  { tokenName: "Wrapped Bitcoin", tokenSymbol: "WBTC", collateralDecimals: 8 }
+  // { tokenSymbol: "WETH", collateralDecimals: 18, syntheticDecimals: 18, priceFeedDecimals: 18 },
+  // { tokenSymbol: "BTC", collateralDecimals: 8, syntheticDecimals: 18, priceFeedDecimals: 8 },
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/financial-templates-lib/test/clients/ExpiringMultiPartyEventClient.js:24]: nope. thanks for pointing this out!

File: packages/financial-templates-lib/src/clients/ExpiringMultiPartyEventClient.js [hunk 8]
```diff
@@ -277,17 +304,27 @@ class ExpiringMultiPartyEventClient {
...
+    // need to be updated in future if we need access to more event info.
     for (let event of liquidationWithdrawnEventsObj) {
       this.liquidationWithdrawnEvents.push({
         transactionHash: event.transactionHash,
         blockNumber: event.blockNumber,
         caller: event.returnValues.caller,
-        withdrawalAmount: event.returnValues.withdrawalAmount,
+        // TODO: refactor this method to use > syntax for versions. this will require a bit of an overhaul to enable version comparison.
+        withdrawalAmount:
+          this.contractVersion == "1.2.0" || this.contractVersion == "1.2.1" || this.contractVersion == "1.2.2"
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/financial-templates-lib/src/clients/ExpiringMultiPartyEventClient.js:319]: great idea. I've left this for a separate PR. added an issue here to track it: https://github.com/UMAprotocol/protocol/issues/2521

**Pull request merged**